src/wavelet_matrix.rs

```diff
@@ -0,0 +1,559 @@
+use anyhow::{anyhow, Result};
+use std::ops::Range;
+
+use crate::{broadword, BitVector, RsBitVector};
+
+/// Time and space efficient index data structures for sequence.
+///
+/// [`WaveletMatrix`] implements stores a sequence of integers and provides myriad operations on sequece.
+/// When a sequence stores $`n`$ integers from $`[0, u-1]`$,
+/// most of the operations run in $`O(log u)`$ , using  $`n log u + o (n log u) + O(log u log n)`$ bits.
+///
+/// This is a yet another Rust port of [hillbig's waveletMatrix](https://github.com/hillbig/waveletTree/blob/master/waveletMatrix.go).
+///
+/// # Example
+///
+/// ```
+/// use sucds::{WaveletMatrix, WaveletMatrixBuilder};
+///
+/// let mut wmb = WaveletMatrixBuilder::new();
+/// let s = "tobeornottobethatisthequestion";
+/// s.chars().for_each(|c| wmb.push(c as usize));
+///
+/// let wm = wmb.build().unwrap();
+/// assert_eq!(wm.num(), s.len());
+/// assert_eq!(wm.dim(), ('u' as usize) + 1);
+///
+/// assert_eq!(wm.lookup(20), 'h' as usize);
+/// assert_eq!(wm.rank(22, 'o' as usize), 4);
+/// assert_eq!(wm.select(2, 't' as usize), 9);
+/// ```
+///
+/// # References
+///
+/// - F. Claude, and G. Navarro, "The Wavelet Matrix," In SPIRE 2012.
+#[derive(Default, Debug, PartialEq, Eq)]
+pub struct WaveletMatrix {
+    layers: Vec<RsBitVector>,
+    dim: usize,
+    num: usize,
+    bit_length: usize,
+}
+
+impl WaveletMatrix {
+    /// Creates a new [`WaveletMatrix`].
+    ///
+    /// # Arguments
+    ///
+    /// - `layers`: A sequence of `RsBitVector`.
+    /// - `dim`: The maximum value + 1 in stored integers.
+    /// - `num`: The number of integers to be stored.
+    /// - `bit_length `: The number of bits needed for storing input integers, i.e. $`log dim`$.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use sucds::{WaveletMatrix, WaveletMatrixBuilder};
+    ///
+    /// let mut wmb = WaveletMatrixBuilder::new();
+    /// let s = "tobeornottobethatisthequestion";
+    /// s.chars().for_each(|c| wmb.push(c as usize));
+    ///
+    /// let wm = wmb.build().unwrap();
+    /// assert_eq!(wm.num(), s.len());
+    /// assert_eq!(wm.dim(), ('u' as usize) + 1);
+    ///
+    /// assert_eq!(wm.lookup(20), 'h' as usize);
+    /// assert_eq!(wm.rank(22, 'o' as usize), 4);
+    /// assert_eq!(wm.select(2, 't' as usize), 9);
+    /// ```
+    pub fn new(layers: Vec<RsBitVector>, dim: usize, num: usize, bit_length: usize) -> Self {
+        Self {
+            layers,
+            dim,
+            num,
+            bit_length,
+        }
+    }
```

Comment:
```suggestion
```

I don't think this interface is necessary, as it makes it difficult for user to understand the arguments and requires additional error handling for invalid input.

Reply:
fixed in 369c984